libusb/descriptor.c

```diff
@@ -1270,12 +1270,12 @@ static int parse_iad_array(struct libusb_context *ctx,
 		iad_array->iad = iad;
 
 		// Second pass: Iterate through desc list, fill IAD structures
-		consumed = 0;
+		int remaining = size;
 		i = 0;
-		while (consumed < size) {
+		do {
 			header.bLength = buffer[0];
 			header.bDescriptorType = buffer[1];
-			if (header.bDescriptorType == LIBUSB_DT_INTERFACE_ASSOCIATION) {
+			if (header.bDescriptorType == LIBUSB_DT_INTERFACE_ASSOCIATION && (remaining >= 8)) {
```

Comment:
Why not sizeof() here as well, instead of "magic" 8?

Reply:
I took a look, but not sure what I could substitute... the 8 is because the next lines read elements 0 through 7.  Using `sizeof libusb_interface_association_descriptor` doesn't feel right, because that's the destination side, not the source side.  And also that struct is not `LIBUSB_PACKED`.

Reply:
If it is clear it should be 8, we can still write 8 instead of an ever so correct, formal compiler abstraction, but we could consider a small comment like a valid IAD descriptor must have 8 bytes. We also define many descriptor length macros in libusb.h, maybe that is over the top here. I guess I got hung up in the inconsistency where above we use a sizeof instead of 2 (I see it more like any descriptor must be long enough to contain its own header otherwise we cannot localize the next, rather than having to fill up a struct if copied).

Reply:
`sizeof (libusb_interface_association_descriptor)` is tempting but incorrect, because a compiler is free to add padding to a struct.

I like the idea of a new constant, see new updated commit.  How shall we name/spell them?

Reply:
For consistency it should be the name of the descriptor type macro + `_SIZE` so LIBUSB_DT_INTERFACE_ASSOCIATION_SIZE even if it's long. DT means descriptor type, so LIBUSB_DT_HEADER doesn't rhyme since it is not specific to a type. We can make something like a  LIBUSB_SMALLEST_DESCRIPTOR_SIZE and set it to 2, or even 3, unless there is a data-less descriptor type of size = 2 (unknown to me). Theoretically even a type-less descriptor of size = 1 would be parseable but probably not something that we should accept as valid input. Unless the same comparison is done elsewhere, so the macro can be reused, maybe it is better to just write 2 and leave a comment about why?

Reply:
Done.

Reply:
I've updated the commit message on the last commit, notably removing the WIP.

Reply:
We already use DESC_HEADER_LENGTH for exactly this in other places in the same file so no need for LIBUSB_SMALLEST_DESCRIPTOR_SIZE.

Reply:
Done.